src/properties/forest.c

```diff
@@ -0,0 +1,215 @@
+    igraph_adjlist_t al;
+    igraph_vector_t visited;
+    igraph_integer_t visited_count=0;
+    igraph_integer_t vcount, ecount;
+
+    vcount = igraph_vcount(graph);
+    ecount = igraph_ecount(graph);
+
+    /*By convention, a zero-vertex graph will be considered a forest.*/
+    if (vcount == 0) {
+        *res = 1;
+        return IGRAPH_SUCCESS;
+    }
+    // A forest can have maximum vcount-1 edges.
+    if (ecount > vcount - 1) {
+        *res = 0;
+        return IGRAPH_SUCCESS;
+    }
+    /*A single-vertex graph is a forest, provided it has no edges (checked in the previous if (..)) */
+    if (vcount == 1) {
+        *res = 1;
+        igraph_vector_push_back(roots,0);
+        return IGRAPH_SUCCESS;
+    }
+    /* Ignore mode for undirected graphs. */
+    if (! igraph_is_directed(graph)) {
+        mode = IGRAPH_ALL;
+    }
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &al, mode, IGRAPH_LOOPS, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+
+    *res = 1; /* assume success */
+    IGRAPH_CHECK(igraph_vector_init(&visited, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &visited);
+
+    IGRAPH_CHECK(igraph_vector_resize(&visited,vcount));
+    igraph_integer_t i;
+    for(i =0; i<vcount;++i){
+        VECTOR(visited)[i]=0;
+    }
+
+    /* The main algorithm:
+     * Undirected Graph:- We add each unvisited vertex to the roots vector, and
+     * mark all other vertices that are reachable from it as visited.
+     *
+     * Directed Graph:- For each tree, the root is the node with no
+     * incoming/outgoing connections, depending on 'mode'. We add each vertex
+     * with zero degree to the roots vector and mark all other vertices that are
+     * reachable from it as visited.
+     *
+     * If all the vertices are visited exactly once, then the graph is a forest.
+     */
+
+    switch (mode) {
+    case IGRAPH_ALL:{
+
+        for(i =0; i<vcount;++i){
+            if (!VECTOR(visited)[i]) {
+                igraph_vector_push_back(roots,i);
+                IGRAPH_CHECK(igraph_i_is_forest_visitor(i,&visited, &al, &visited_count,res,mode));
+            }
+        }
+        break;
+    }
+    case IGRAPH_IN:
+    case IGRAPH_OUT: {
+        igraph_vector_t degree;
+
+        IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+        IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+        IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), mode == IGRAPH_IN ? IGRAPH_OUT : IGRAPH_IN, /* loops = */ 1));
+
+        for (i = 0; i < vcount; ++i){
+            if (VECTOR(degree)[i] == 0) {
+                igraph_vector_push_back(roots,i);
+                IGRAPH_CHECK(igraph_i_is_forest_visitor(i,&visited, &al, &visited_count,res,mode));
+            }
+        }
+
+        igraph_vector_destroy(&degree);
+        IGRAPH_FINALLY_CLEAN(1);
+        break;
+    }
+    default:
+        IGRAPH_ERROR("Invalid mode", IGRAPH_EINVMODE);
+    }
+    if(*res){
+        *res= visited_count==vcount;
```

Comment:
Can you give me an example of when then visited_count is not the vcount? I think that shouldn't happen for this algorithm.

Reply:
If there is a single vertex in directed graph with a self loop then its degree wont be 0 so `igraph_forest_visitor` wont be called for that vertex and function would return that it is a forest.

Reply:
Aha! Thanks!